Read SoA scalars and Eigen columns from a ROOT streamer

fwyzard opened the pull request:
#### PR description:

Fix the ROOT streamers for the template-based SoA to read back the scalar variables, and to write and read back Eigen objects.

Update the `TestAlpakaAnalyzer` and CUDA `TestPortableAnalyzer` to check that scalars and matrices are read back correctly, and print a more verbose description of the SoA being read.

Update the test scripts so that the unit tests actually fail (exit with a non-zero value) if any of the cmsRun jobs fails.

#### PR validation:

The improved unit tests pass.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

To be backported to CMSSW_12_5_X for the Alpaka migration.

makortel committed:
Add missing includes

fwyzard committed:
Make the TestAlpakaAnalyzer and CUDA TestPortableAnalyzer more verbose

fwyzard committed:
Make the CUDA and Alpaka write/read tests fail on errors

fwyzard committed:
Implement the ROOT streamer for reading back SoA scalars

fwyzard committed:
Check that scalars are set and read back correctly

Move the printout before the assert, so it can provide more details if
the assertion fails.

fwyzard commented:
type bugfix

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39532/32305

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaCore/interface/chooseDevice.h  modified in PR(s): #39498

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- DataFormats/SoATemplate (**heterogeneous**)
- HeterogeneousCore/AlpakaCore (**heterogeneous**)
- HeterogeneousCore/AlpakaTest (**heterogeneous**)


@makortel, @fwyzard can you please review it and eventually sign? Thanks.
@makortel, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ericcano committed:
Add ROOT streaming support for Eigen SoA columns

Store Eigen SoA columns as a strided, linearised array, and implement a
ROOT streamer for reading them back as Eigen objects in SoA columns.

Update the CUDA and Alpaka portable collection tests to fill the Eigen
Matrix column, print part of its content, and check that it is written
and read back correctly.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-474471/27840/summary.html
**COMMIT**: 848368c517acd061939380c6945e92120bb8577e
**CMSSW**: CMSSW_12_6_X_2022-09-28-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39532/27840/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testHeterogeneousCoreAlpakaTestWriteRead had ERRORS
</pre>

## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-474471/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3433154
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3433129
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-28-2300+474471/53027/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - Reco comparison had  3 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 74
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19802
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_6_X_2022-09-28-2300+474471/53024/triggerResults): no differences found

makortel commented on DataFormats/SoATemplate/interface/SoALayout.h:
```diff
@@ -361,13 +361,13 @@
       ,                                                                                                                \
       /* Eigen column */                                                                                               \
-      /* TODO: implement*/                                                                                             \
+      std::cerr << "Warning: reading an Eigen column from a ROOT file is not implemented" << std::endl;                \
```
Would an exception make sense? (or does it really matter for now?)

fwyzard replied:
@ericcano should have an implementation of it later today, I'll add it here and update the PR description.

makortel reviewed (commented):
The first commit (e5736f4e9a11bbf5a4fbb52e358463af8a348c0e) is already part of the `master` branch. Can it be dropped?

fwyzard commented:
test parameters:
  - cms-addpkg: DataFormats/SoATemplate, DataFormats/Portable, DataFormats/PortableTestObjects, HeterogeneousCore/AlpakaCore, HeterogeneousCore/AlpakaTest
  - enable_tests: gpu

fwyzard commented:
test parameters:
  - cms-addpkg: DataFormats/SoATemplate,DataFormats/PortableTestObjects,HeterogeneousCore/AlpakaCore,HeterogeneousCore/AlpakaTest
  - enable_test: gpu

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39532/32314

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaCore/interface/chooseDevice.h  modified in PR(s): #39498

cmsbuild commented:
Pull request #39532 was updated. @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39532/32317

ERROR: Build errors found during clang-tidy run.
```
                                             typename 
--
DataFormats/SoATemplate/interface/SoACommon.h:506:46: error: missing 'typename' prior to dependent type name 'Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
DataFormats/SoATemplate/interface/SoACommon.h:506:46: error: missing 'typename' prior to dependent type name 'const Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
DataFormats/SoATemplate/interface/SoACommon.h:506:46: error: missing 'typename' prior to dependent type name 'Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
DataFormats/SoATemplate/interface/SoACommon.h:506:46: error: missing 'typename' prior to dependent type name 'const Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

makortel commented:
> Looks like SCRAM is not checking out and recompiling all the dependencies (or we failed to declare some of them ?).

I see
https://github.com/cms-sw/cmssw/blob/bdd7ac01d973e46f95f1bf4182ae65d440beb497/DataFormats/PortableTestObjects/BuildFile.xml#L1-L8
is missing a dependence on `DataFormats/SoATemplate` caused by
https://github.com/cms-sw/cmssw/blob/bdd7ac01d973e46f95f1bf4182ae65d440beb497/DataFormats/PortableTestObjects/interface/TestSoA.h#L4-L6

fwyzard committed:
Add missing dependency

fwyzard commented:
Thanks, let's see if adding it helps !

fwyzard commented:
enable gpu

fwyzard commented:
please test

fwyzard committed:
Add missing typename specifier

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39532/32321

ERROR: Build errors found during clang-tidy run.
```
DataFormats/SoATemplate/interface/SoACommon.h:393:46: error: missing 'typename' prior to dependent type name 'Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
DataFormats/SoATemplate/interface/SoACommon.h:393:46: error: missing 'typename' prior to dependent type name 'Matrix<double, 3, 6, 0>::Scalar' [clang-diagnostic-error]
    static constexpr auto valueSize = sizeof(C::Scalar);
                                             ^
                                             typename 
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

fwyzard commented:
test parameters:
  - enable_tests: gpu

fwyzard commented:
please test

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39532/32327

- This PR adds an extra 36KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaCore/interface/chooseDevice.h  modified in PR(s): #39498

cmsbuild commented:
Pull request #39532 was updated. @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
please test with cmsdist#8107

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-474471/27868/summary.html
**COMMIT**: 954edbc5dedbc62d6e1e10acfbaced01d2116b7f
**CMSSW**: CMSSW_12_6_X_2022-09-29-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39532/27868/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /pool/condor/dir_268586/jenkins/workspace/compare-root-files-short-matrix/data/PR-474471/41834.0_TTbar_14TeV+2026D94+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 19 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3433154
 - DQMHistoTests: Total failures: 114
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3433018
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 49 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-09-29-2300+474471/53054/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - Reco comparison had  3 failed jobs
 - DQMHistoTests: Total files compared: 4
 - DQMHistoTests: Total histograms compared: 19876
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 19868
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 3 files compared)
 - Checked 12 log files, 9 edm output root files, 4 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_12_6_X_2022-09-29-2300+474471/53053/triggerResults): no differences found

fwyzard commented:
@makortel do you have any other concerns, or is it ok to merge this ?

makortel commented:
> @makortel do you have any other concerns, or is it ok to merge this ?

Looks good, although I would still prefer to drop the first commit (e5736f4e9a11bbf5a4fbb52e358463af8a348c0e) since it is already part of the master branch (even if this time git/GitHub is able to see it doesn't cause conflict).

fwyzard commented:
That's the trick I found (by chance) so that all future PRs based on this same branch will merge without conflicts in both 12.5.x and 12.6.x.

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
@fwyzard you tested this lately together with https://github.com/cms-sw/cmsdist/pull/8107, but neither here nor in the description of that cmsdist PR is ever written that the two havo to be merged together: please confirm that this PR can be merged independently from https://github.com/cms-sw/cmsdist/pull/8107, or provide some different direction otherwise.

fwyzard commented:
@perrotta, technically this and 8107 are independent.

8107 is a fix that can go in by itself.

This PR implements changes that are independent from 8107, and (also) makes the unit test report any failures instead of ignoring them.

Some random failures are already happening, but without this PR they are ignored: 8107 fixes them.

perrotta commented:
+1

perrotta commented:
@fwyzard perhaps you already noticed it, but after the merge of this PR the unit test in HeterogeneousCore/AlpakaTest/test/testHeterogeneousCoreAlpakaTestWriteRead.sh started failing, see e.g. https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc10/CMSSW_12_6_X_2022-10-01-1100/unitTestLogs/HeterogeneousCore/AlpakaTest#/

fwyzard commented:
Yes - the fix is cms-sw/cmsdist#8107 .

fwyzard commented:
What I meant [earlier](https://github.com/cms-sw/cmssw/pull/39532#issuecomment-1264277732) is that the underlying tests has been failing for a while (and cms-sw/cmsdist#8107 is the fix).

However, the failures were going unreported due to an [annoying interplay](https://mywiki.wooledge.org/BashFAQ/105) between how scram run the tests and how bash inherits an environment from a parent process running within a conditional statement. This PR [fixes the scripts to actually report the errors](https://github.com/cms-sw/cmssw/pull/39532/files#diff-4d9d99f0277ca0596a6dbcabc019e32068a2886437f2a7edf7625db6cd553d12) - so the pre-existing failures are now apparent.
